### DISCUSSION TITLE
RequestValidationError and ValidationError under Pydantic V2

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [pydantic/pydantic#7107] [Pydantic V2]
I posted some of this on Pydantic discussions but the question belongs here. I have a FastApi application (w/ Pydantic V1) and want to migrate it to FastApi v0.101.0+ (with Pydantic V2). My (V1)  application error handling successfully uses this construct:
```
from fastapi.exceptions import HTTPException, RequestValidationError
from pydantic import ErrorWrapper

try:
    this = something('odd')
except:
    raise RequestValidationError(errors=[ErrorWrapper(ValueError('Invalid'), 'odd')])
```
With FastApi v0.101 and Pydantic V2, and replacing ErrorWrapper with ValidationError, the above code causes a TypeError:
```
from fastapi.exceptions import HTTPException, RequestValidationError
from pydantic import ValidationError

try:
    this = something('odd')
except:
    raise RequestValidationError(errors=[ValidationError(ValueError('Invalid'), 'odd')])

TypeError: No constructor defined
```

The Pydantic suggestion is to use ValidationError.from_exception_data like this (I think), which also causes a TypeError:
```
from fastapi.exceptions import HTTPException, RequestValidationError
from pydantic import ValidationError

try:
    this = something('odd')
except:
    raise RequestValidationError(errors=[ValidationError.from_exception_data(ValueError('Invalid'), 'odd')])

TypeError: argument 'line_errors': 'str' object cannot be converted to 'PyList'
```
I use mypy for type checking and have also looked at https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core._pydantic_core.ValidationError.from_exception_data but am none the wiser :(

What am I not understanding and what is the way forward? Thanks

**Comment by Viicos** [comment]
A `ValidationError` is specific to pydantic validation, if your call to `something('odd')` has nothing to do with data validation, you shouldn't use it (same thing goes for `RequestValidationError`, if it has nothing to do with the incoming request it shouldn't be used).

If you still want to use `ValidationError`, the signature is as follows:
```python
    @staticmethod
    def from_exception_data(
        title: str,
        line_errors: list[InitErrorDetails],
        error_mode: Literal['python', 'json'] = 'python',
        hide_input: bool = False,
    ) -> ValidationError:
```

and `InitErrorDetails` is a typed dict:

```python
class InitErrorDetails(TypedDict):
    type: str | PydanticCustomError
    """The type of error that occurred, this should a "slug" identifier that changes rarely or never."""
    loc: _NotRequired[tuple[int | str, ...]]
    """Tuple of strings and ints identifying where in the schema the error occurred."""
    input: _Any
    """The input data at this `loc` that caused the error."""
    ctx: _NotRequired[dict[str, str | int | float]]
    """
    Values which are required to render the error message, and could hence be useful in rendering custom error messages.
    """
```

As you can see, it is specifically made for pydantic validation

**Comment by dineshbvadhia** [comment]
Ooops ... I meant to post this on the FastAPi discussion site.

Thank-you for the response. Let's see what the FastApi folks have to say.
